src/dbnode/storage/index/compaction/plan.go

```diff
@@ -0,0 +1,278 @@
+			continue
+		}
+		// we need to compact mutable segments regardless of whether they belong to a known level.
+		if seg.Type == segments.MutableType {
+			catchAllMutableSegmentTask.Segments = append(catchAllMutableSegmentTask.Segments, seg)
+			continue
+		}
+		// in all other situations, we simply mark the segment unused and move on
+		plan.UnusedSegments = append(plan.UnusedSegments, seg)
+	}
+
+	// any segments that don't fit any known buckets
+	if len(catchAllMutableSegmentTask.Segments) != 0 {
+		plan.Tasks = append(plan.Tasks, Task{
+			Segments: catchAllMutableSegmentTask.Segments,
+		})
+	}
+
+	// for each level, sub-group segments into tier'd sizes (b)
+	for level, levelSegments := range segementsByLevel {
+		var (
+			task            Task
+			accumulatedSize int64
+		)
+		sort.Slice(levelSegments, func(i, j int) bool {
+			return levelSegments[i].Size < levelSegments[j].Size
+		})
+		for _, seg := range levelSegments {
+			accumulatedSize += seg.Size
+			task.Segments = append(task.Segments, seg)
+			if accumulatedSize >= level.MaxSizeExclusive {
+				plan.Tasks = append(plan.Tasks, task)
+				task = Task{}
+				accumulatedSize = 0
+			}
+		}
+		// fall thru cases: no accumulation, so we're good
+		if len(task.Segments) == 0 || accumulatedSize == 0 {
+			continue
+		}
+
+		// in case we never went over accumulated size, but have 2 or more segments, we should still compact them
+		if len(task.Segments) > 1 {
+			plan.Tasks = append(plan.Tasks, task)
+			continue
+		}
+
+		// even if we only have a single segment, if its a mutable segment, we should compact it to convert into a FST
+		if task.Segments[0].Type == segments.MutableType {
+			plan.Tasks = append(plan.Tasks, task)
+			continue
+		}
+
+		// at this point, we have a single FST segment but don't need to compact it; so mark it as such
+		plan.UnusedSegments = append(plan.UnusedSegments, task.Segments[0])
+	}
+
+	// now that we have the plan, we priortise the tasks as requested in the opts. (c)
+	sort.Stable(plan)
+	return plan, nil
+}
+
+func (p *Plan) Len() int      { return len(p.Tasks) }
+func (p *Plan) Swap(i, j int) { p.Tasks[i], p.Tasks[j] = p.Tasks[j], p.Tasks[i] }
+func (p *Plan) Less(i, j int) bool {
+	switch p.OrderBy {
+	case TasksOrderedByOldestMutableAndSize:
+		fallthrough
+	default:
+		// NB: the intent with the conditions below is to optimise for e2e ingest latency first,
+		// which is why we prefer to compact older mutable segments first, then any larger ones,
+		// after which, we fall back to the graceful plan of compacting smaller segments over
+		// larger ones to reduce total compactions required.
+		taskSummaryi, taskSummaryj := p.Tasks[i].Summary(), p.Tasks[j].Summary()
+		if taskSummaryi.CumulativeMutableAge != taskSummaryj.CumulativeMutableAge {
+			// i.e. put those tasks which have cumulative age greater first
+			return taskSummaryi.CumulativeMutableAge > taskSummaryj.CumulativeMutableAge
```

Comment:
Hm, this method should return true if `i < j` no?  Should this not be `return taskSummaryi.CumulativeMutableAge M taskSummaryj.CumulativeMutableAge`?